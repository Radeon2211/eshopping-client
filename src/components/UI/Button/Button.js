import React from 'react';
import PropTypes from 'prop-types';
import * as SC from './Button.sc';

const Button = (props) => {
  const { clicked, children } = props;
  return (
    <SC.Button onClick={clicked} {...props}>
      {children}
    </SC.Button>
  );
};

Button.defaultProps = {
  type: 'button',
  filled: false,
  disabled: false,
  clicked: () => {},
};

Button.propTypes = {
  type: PropTypes.string,
  size: PropTypes.string.isRequired,
  filled: PropTypes.bool,
  disabled: PropTypes.bool,
  clicked: PropTypes.func,
  children: PropTypes.node.isRequired,
};

export default Button;
