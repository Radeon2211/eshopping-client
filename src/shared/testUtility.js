// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const historyPageNum = (pageNumber) => ({
  listen: jest.fn(),
  location: { search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push: jest.fn(),
});

export const propsPagination = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  maxQuantityPerPage: 2,
});
