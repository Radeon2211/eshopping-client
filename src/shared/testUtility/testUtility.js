// eslint-disable-next-line import/no-extraneous-dependencies
import checkPropTypes from 'check-prop-types';
import { v4 as uuidv4 } from 'uuid';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../../store/reducers/rootReducer';
import { productConditions } from '../constants';

export const defaultDeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: 'active',
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const checkProps = (component, expectedProps) => {
  // eslint-disable-next-line react/forbid-foreign-prop-types
  return checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

export const createHistoryPageNumber = (pageNumber = 1, push = jest.fn()) => ({
  listen: jest.fn(),
  location: { pathname: '/products', search: `?p=${pageNumber}` },
  createHref: jest.fn(),
  push,
});

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
  quantitySold = 0,
  buyerQuantity = 0,
  description = '',
  condition = productConditions.NEW,
) => ({
  _id: productId || uuidv4(),
  name,
  price,
  quantity,
  quantitySold,
  buyerQuantity,
  description,
  condition,
  photo,
  seller: {
    username: sellerUsername,
  },
  __v: 0,
  createdAt: '2021-02-10T19:10:38.872Z',
  updatedAt: '2021-02-10T19:10:38.872Z',
});

export const createCartItem = (
  sellerUsername = 'username',
  quantity = 1,
  productId,
  price = 2,
  name = 'product name',
  productQuantity = 2,
  photo = false,
) => {
  const product = createProductItem(
    productId || uuidv4(),
    sellerUsername,
    productQuantity,
    price,
    name,
    photo,
  );

  return {
    _id: uuidv4(),
    quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  productId,
  sellerUsername = 'username',
  quantity = 1,
  price = 2,
  name = 'product name',
  photo = false,
) => ({
  _id: productId || uuidv4(),
  name,
  photo,
  price,
  quantity,
  seller: {
    username: sellerUsername,
  },
});

export const createOrder = (
  products = [createTransactionAndOrderProdItem()],
  orderId = uuidv4(),
  sellerUsername,
  buyerUsername,
  overallPrice,
  createdAt,
  sellerEmail,
  sellerPhone,
) => {
  const seller = sellerUsername
    ? {
        username: sellerUsername,
        email: sellerEmail,
        phone: sellerPhone,
      }
    : null;

  const buyer = buyerUsername
    ? {
        username: buyerUsername,
      }
    : null;

  return {
    _id: orderId,
    seller,
    buyer,
    products,
    overallPrice,
    deliveryAddress: defaultDeliveryAddress,
    createdAt,
    updatedAt: createdAt,
    __v: 0,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...defaultState.product,
      ...initialProduct,
    },
    ui: {
      ...defaultState.ui,
      ...initialUI,
    },
  };
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

export const checkReqMethodAndURL = (moxiosInstance, method, url, reqNumber = 0) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests[reqNumber - 1];
    return method === prevRequest.config.method && url === prevRequest.config.url;
  }
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState,
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};
