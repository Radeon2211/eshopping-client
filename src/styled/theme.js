const theme = {
  colors: {
    light1: '#FAFAFA',
    light1Transparent: '#FAFAFA88',
    light2: '#DADFE1',
    light3: '#BFBFBF',
    light4: '#666',
    blue: '#1E8BC3',
    blueLight: '#22A7F0',
    blueDark: '#67809F',
    green: '#00A790',
    greenDark: '#009680',
    red: '#F22613',
  },
  shadows: {
    level1: '0 0 2px rgba(0, 0, 0, .1)',
    level2: '0 0 2px rgba(0, 0, 0, .3)',
    level3: '0 0 3px rgba(0, 0, 0, .3)',
    level4: '0 0 5px rgba(0, 0, 0, .3)',
    level5: '0 0 8px rgba(0, 0, 0, .3)',
  },
  fonts: {
    logo: 'Lobster, cursive',
    heading: 'Ubuntu, sans-serif',
    text: 'Lato, sans-serif',
  },
  spacings: {
    level1: '0.8rem',
    level2: '1.2rem',
    level3: '2.4rem',
    level4: '3.6rem',
    level5: '4.8rem',
  },
  zIndexes: {
    level1: 10,
    level2: 100,
    level3: 1000,
    level4: 10000,
  },
};

export default theme;
