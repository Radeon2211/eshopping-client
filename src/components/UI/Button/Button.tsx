import * as SC from './Button.sc';
import LoadingOverlay from '../LoadingOverlay/LoadingOverlay';
import { PropsWithChildren } from '../../../shared/types/types';

export interface ButtonProps {
  $size?: 'small' | 'big';
  type?: 'submit' | 'reset' | 'button';
  $color?: 'blue' | 'red';
  $filled?: boolean;
  $stretch?: boolean;
  isLoading?: boolean;
  disabled?: boolean;
  clicked?: () => void;
}

export default function Button(props: PropsWithChildren<ButtonProps>) {
  const {
    clicked = () => {},
    isLoading = false,
    disabled = false,
    type = 'button',
    $size = 'small',
    $color = 'blue',
    $stretch = false,
    $filled = false,
    children,
    ...rest
  } = props;

  const loadingOverlay = isLoading ? <LoadingOverlay loaderSize="small" disableText /> : null;

  return (
    <SC.Button
      onClick={clicked}
      disabled={disabled || isLoading}
      type={type}
      $size={$size}
      $color={$color}
      $stretch={$stretch}
      $filled={$filled}
      {...rest}
    >
      {children}
      {loadingOverlay}
    </SC.Button>
  );
}
