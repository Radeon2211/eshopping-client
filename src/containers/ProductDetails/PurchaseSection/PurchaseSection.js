import React, { useState, useCallback } from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import { Link, useHistory } from 'react-router-dom';
import * as SC from './PurchaseSection.sc';
import * as actions from '../../../store/actions/indexActions';
import { modalTypes } from '../../../shared/constants';
import Button from '../../../components/UI/Button/Button';
import ChooseQuantity from '../../../components/UI/ChooseQuantity';
import { GreenText, GrayText } from '../../../styled/components';
import Heading from '../../../components/UI/Heading/Heading';

const PurchaseSection = (props) => {
  const { productId, productQuantity, productSellerUsername, onSetModal, userProfile } = props;

  const history = useHistory();

  const [chosenQuantity, setChosenQuantity] = useState(1);

  const cart = useSelector((state) => state.auth.cart);
  const isCartLoading = useSelector((state) => state.ui.isCartLoading);

  const dispatch = useDispatch();
  const onAddCartItem = useCallback((item) => dispatch(actions.addCartItem(item)), [dispatch]);
  const onGoToTransaction = useCallback(
    (currentHistory, item) => dispatch(actions.goToTransaction(currentHistory, item)),
    [dispatch],
  );

  const inputChangeHandle = (e) => {
    const value = +e.target.value || '';
    if (!value) {
      setChosenQuantity(value);
      return;
    }
    if (value < 1 || value > productQuantity) return;
    setChosenQuantity(value);
  };

  const inputBlurHandle = (e) => {
    const value = +e.target.value;
    if (!value || value < 1) {
      setChosenQuantity(1);
    } else if (value > productQuantity) {
      setChosenQuantity(productQuantity);
    } else {
      setChosenQuantity(value);
    }
  };

  const incrementQuantityHandle = () => {
    if (chosenQuantity < productQuantity) {
      setChosenQuantity((prevState) => prevState + 1);
    }
  };

  const decrementQuantityHandle = () => {
    if (chosenQuantity > 1) {
      setChosenQuantity((prevState) => prevState - 1);
    }
  };

  const addToCartClickHandle = () => {
    if (!userProfile) {
      onSetModal(true, modalTypes.LOGIN);
    } else {
      onAddCartItem({
        quantity: chosenQuantity,
        product: productId,
      });
    }
  };

  const buyNowClickHandle = () => {
    if (!userProfile) {
      onSetModal(true, modalTypes.LOGIN);
    } else {
      onGoToTransaction(history, {
        product: productId,
        quantity: chosenQuantity,
      });
    }
  };

  let purchaseSection = (
    <>
      <span className="quantity-info gray-text">{`Quantity: ${productQuantity}`}</span>
      <Heading variant="h4" mgTop="level3" data-test="info-to-seller">
        You are the seller of this product
      </Heading>
    </>
  );

  if (userProfile?.username !== productSellerUsername) {
    let givenProductInCart = null;
    if (cart) {
      givenProductInCart = cart.find((item) => item.product._id === productId);
    }
    let addToCartBtn = (
      <Button filled stretch clicked={addToCartClickHandle} isLoading={isCartLoading}>
        Add to cart
      </Button>
    );
    if (givenProductInCart?.quantity >= productQuantity) {
      addToCartBtn = (
        <Heading variant="h4" align="center" mgBottom="level2" data-test="not-able-to-add">
          You have added all pieces to&nbsp;
          <Link to="/cart">
            <GreenText>cart</GreenText>
          </Link>
        </Heading>
      );
    }

    purchaseSection = (
      <>
        <div className="choose-quantity-box">
          <ChooseQuantity
            name="quantity"
            maxQuantity={productQuantity}
            value={chosenQuantity}
            incremented={incrementQuantityHandle}
            decremented={decrementQuantityHandle}
            changed={inputChangeHandle}
            blured={inputBlurHandle}
          />
          <span className="quantity-number">
            {`of ${productQuantity} piece${productQuantity > 1 ? 's' : ''}`}
            <GrayText>
              {givenProductInCart ? ` (${givenProductInCart?.quantity} in cart)` : ''}
            </GrayText>
          </span>
        </div>
        {addToCartBtn}
        <Button filled stretch clicked={buyNowClickHandle} isLoading={isCartLoading}>
          buy now
        </Button>
      </>
    );
  }

  return <SC.Wrapper>{purchaseSection}</SC.Wrapper>;
};

PurchaseSection.defaultProps = {
  productSellerUsername: undefined,
  userProfile: null,
};

PurchaseSection.propTypes = {
  productId: PropTypes.string.isRequired,
  productQuantity: PropTypes.number.isRequired,
  productSellerUsername: PropTypes.string,
  onSetModal: PropTypes.func.isRequired,
  userProfile: PropTypes.oneOfType([PropTypes.object]),
};

export default PurchaseSection;
