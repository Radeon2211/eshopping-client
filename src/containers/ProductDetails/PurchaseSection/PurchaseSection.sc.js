import styled from 'styled-components';

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-items: stretch;

  & > button:not(:last-child) {
    margin-bottom: ${({ theme }) => theme.spacings.level2};
  }

  & .info-quantity-box {
    margin-bottom: ${({ theme }) => theme.spacings.level2};
  }

  & .choose-quantity-box {
    align-items: center;
    margin-bottom: ${({ theme }) => theme.spacings.level3};
    display: flex;
  }

  & .quantity-number {
    font-size: ${({ theme }) => theme.fontSizes.level2};
    margin-left: ${({ theme }) => theme.spacings.level1};
  }

  & .not-able-to-add {
    font-size: ${({ theme }) => theme.fontSizes.level4};
    font-weight: 700;
    margin-bottom: ${({ theme }) => theme.spacings.level2};
    text-align: center;
  }
`;

export const InfoToSeller = styled.div`
  display: flex;
  flex-direction: column;

  & .quantity-box {
    margin-bottom: ${({ theme }) => theme.spacings.level2};
  }

  & .quantity-text {
    font-size: ${({ theme }) => theme.fontSizes.level2};
  }

  & .text {
    font-size: ${({ theme }) => theme.fontSizes.level4};
    font-weight: 700;
  }
`;
