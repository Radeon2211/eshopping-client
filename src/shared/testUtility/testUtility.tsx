/* eslint-disable import/no-extraneous-dependencies */
import { ReactElement } from 'react';
import { Router } from 'react-router-dom';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import { MockStoreEnhanced } from 'redux-mock-store';
import { fireEvent, waitFor, render } from '@testing-library/react';
import { v4 as uuidv4 } from 'uuid';
import moxios from 'moxios';
import rootReducer from '../../store/reducers/rootReducer';
import theme from '../../styled/theme';
import {
  CartItem,
  DeliveryAddress,
  Order,
  OrderProductItem,
  Product,
  Profile,
  ProfileStatus,
} from '../types/types';
import { ProductCondition } from '../types/enums';

export const defaultDeliveryAddress: DeliveryAddress = {
  firstName: 'firstName',
  lastName: 'lastName',
  street: 'street',
  zipCode: 'zipCode',
  city: 'city',
  country: 'country',
  phone: 'phone',
};

export const defaultUserProfile: Profile = {
  ...defaultDeliveryAddress,
  _id: 'userId',
  username: 'username',
  email: 'email@domain.com',
  status: ProfileStatus.ACTIVE,
  contacts: {
    email: true,
    phone: true,
  },
  createdAt: '2021-02-10T19:10:38.872Z',
};

export const createPaginationProps = (itemQuantity = 5) => ({
  itemQuantity,
  isDataLoading: false,
  quantityPerPage: 2,
});

export const createProductItem = (props: Partial<Product> = {}): Product => {
  return {
    _id: uuidv4(),
    quantity: 1,
    price: 2,
    name: 'product name',
    photo: false,
    quantitySold: 0,
    buyerQuantity: 0,
    description: '',
    condition: ProductCondition.NEW,
    seller: {
      username: 'username',
    },
    createdAt: '2021-02-10T19:10:38.872Z',
    updatedAt: '2021-02-10T19:10:38.872Z',
    __v: 0,
    ...props,
  };
};

export const createCartItem = (props = {}): CartItem => {
  const finalProps = {
    id: uuidv4(),
    sellerUsername: 'username',
    quantity: 1,
    productId: uuidv4(),
    price: 2,
    name: 'product name',
    productQuantity: 2,
    photo: false,
    ...props,
  };
  const product = createProductItem({
    _id: finalProps.productId,
    quantity: finalProps.productQuantity,
    price: finalProps.price,
    name: finalProps.name,
    photo: finalProps.photo,
    seller: {
      username: finalProps.sellerUsername,
    },
  });

  return {
    _id: finalProps.id,
    quantity: finalProps.quantity,
    product,
  };
};

export const createTransactionAndOrderProdItem = (
  props: Partial<OrderProductItem> = {},
): OrderProductItem => {
  return {
    _id: uuidv4(),
    quantity: 1,
    price: 2,
    name: 'product name',
    photo: false,
    seller: {
      username: 'username',
    },
    ...props,
  };
};

export const createOrder = (props: Partial<Order> = {}): Order => {
  return {
    _id: uuidv4(),
    overallPrice: 100,
    seller: null,
    buyer: null,
    products: [createTransactionAndOrderProdItem()],
    deliveryAddress: defaultDeliveryAddress,
    createdAt: '2021-02-10T19:10:38.872Z',
    updatedAt: '2021-02-10T19:10:38.872Z',
    __v: 0,
    ...props,
  };
};

export const testStore = (initialAuth = {}, initialProduct = {}, initialUI = {}) => {
  const createStoreWithMiddleware = applyMiddleware(thunk)(createStore);
  const defaultState = createStoreWithMiddleware(rootReducer).getState();

  const finalState = {
    auth: {
      ...defaultState.auth,
      ...initialAuth,
    },
    product: {
      ...(defaultState.product as object),
      ...initialProduct,
    },
    ui: {
      ...(defaultState.ui as object),
      ...initialUI,
    },
  };
  // @ts-expect-error TBF after redux upgrade
  const finalStore = createStoreWithMiddleware(rootReducer, finalState);

  return {
    store: finalStore,
    initialState: finalStore.getState(),
  };
};

type MoxiosRequestsItems = { config: { method: string; url: string } }[];
type MockedMoxiosInstance = {
  requests: {
    __items?: MoxiosRequestsItems;
    mostRecent?: () => { config: { method: string; url: string } };
  };
};
type DefaultMoxiosInstance = Omit<typeof moxios, 'requests'> & {
  requests: typeof moxios.requests & {
    __items?: MoxiosRequestsItems;
  };
};
export const checkReqMethodAndURL = (
  moxiosInstance: MockedMoxiosInstance | DefaultMoxiosInstance,
  method: string,
  url: string,
  reqNumber = 0,
) => {
  if (reqNumber) {
    const requests = moxiosInstance.requests.__items;
    const prevRequest = requests?.[reqNumber - 1];
    return prevRequest && method === prevRequest.config.method && url === prevRequest.config.url;
  }
  if (!moxiosInstance.requests.mostRecent) return false;
  const lastRequest = moxiosInstance.requests.mostRecent();
  return method === lastRequest.config.method && url === lastRequest.config.url;
};

export const createExpectedState = (
  initialState: {
    auth?: object;
    product?: object;
    ui?: object;
  },
  expectedAuth = {},
  expectedProduct = {},
  expectedUI = {},
) => {
  const newState = {
    auth: {
      ...initialState.auth,
      ...expectedAuth,
    },
    product: {
      ...initialState.product,
      ...expectedProduct,
    },
    ui: {
      ...initialState.ui,
      ...expectedUI,
    },
  };

  return newState;
};

export const setUpStoreWithDefaultProfile = (auth = {}, product = {}, ui = {}) => {
  return testStore(
    {
      profile: defaultUserProfile,
      deliveryAddress: defaultDeliveryAddress,
      cart: [],
      ...auth,
    },
    product,
    ui,
  );
};

export const clickAtSubmitButton = async (container: Element) => {
  await waitFor(() => {
    const submitBtn = container.querySelector('button[type="submit"]');
    if (submitBtn) {
      fireEvent.click(submitBtn);
    }
  });
};

export const testRouterPushCall = (
  pushFn: ReturnType<typeof jest.fn>,
  callIdx: number,
  expectedPathname: string,
  expectedSearch = '',
) => {
  expect(pushFn.mock.calls[callIdx][0].pathname).toEqual(expectedPathname);
  expect(pushFn.mock.calls[callIdx][0].search).toEqual(expectedSearch);
};

type RenderAppPartOptions = {
  push?: ReturnType<typeof jest.fn>;
  pathname?: string;
  search?: string;
  withoutThemeProvider?: boolean;
  withoutRouter?: boolean;
  store?: MockStoreEnhanced<object, object>;
  onlyReturnWrappedElement?: boolean;
};

export const renderAppPart = (element: ReactElement, options: RenderAppPartOptions = {}) => {
  const navigator = {
    createHref: jest.fn(),
    push: options.push || jest.fn(),
    go: jest.fn(),
    replace: jest.fn(),
  };

  const location = {
    pathname: options.pathname || '/',
    search: options.search || '',
  };

  let content = element;
  if (!options.withoutThemeProvider) {
    content = <ThemeProvider theme={theme}>{element}</ThemeProvider>;
  }
  if (options.store) {
    content = <Provider store={options.store}>{content}</Provider>;
  }
  if (!options.withoutRouter) {
    content = (
      <Router location={location} navigator={navigator}>
        {content}
      </Router>
    );
  }

  return options.onlyReturnWrappedElement ? content : render(content);
};
