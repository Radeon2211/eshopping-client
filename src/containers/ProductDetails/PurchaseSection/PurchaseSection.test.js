import React from 'react';
import { mount } from 'enzyme';
import { HashRouter as Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import PurchaseSection from './PurchaseSection';
import Button from '../../../components/UI/Button/Button';
import { checkProps } from '../../../shared/testUtility';
import theme from '../../../styled/theme';

const defaultProps = {
  onSetModal: jest.fn(),
  userProfile: { _id: '987' },
  productSellerId: '123',
  productQuantity: 5,
};

const setUp = (props) => {
  const finalProps = props
    ? {
        ...defaultProps,
        ...props,
      }
    : defaultProps;
  return mount(
    <Router>
      <ThemeProvider theme={theme}>
        <PurchaseSection {...finalProps} />
      </ThemeProvider>
    </Router>,
  );
};

describe('<PurchaseSection />', () => {
  describe('Check prop types', () => {
    it('Should NOT throw a warning', () => {
      expect(checkProps(PurchaseSection, defaultProps)).toBeUndefined();
    });
    it('Should throw a warning', () => {
      expect(checkProps(PurchaseSection, {})).not.toBe(null);
    });
  });

  describe('Check how overall section renders', () => {
    it('Should render quantity box and two buttons when seller and current user are different', () => {
      const wrapper = setUp();
      expect(wrapper.find('.quantity-box')).toHaveLength(1);
      expect(wrapper.find(Button)).toHaveLength(2);
    });
    it('Should NOT render quantity box and two buttons & render .not-able-to-buy when seller and current user are the same', () => {
      const wrapper = setUp({ userProfile: { _id: '123' } });
      expect(wrapper.find('.quantity-box')).toHaveLength(0);
      expect(wrapper.find(Button)).toHaveLength(0);
      expect(wrapper.find('.not-able-to-buy')).toHaveLength(1);
    });
  });

  describe('Check how quantity number text renders', () => {
    it('Should render appropriate text when quantity is 1', () => {
      const wrapper = setUp({ productQuantity: 1 });
      expect(wrapper.find('.quantity-number').text()).toEqual('of 1 piece');
    });
    it('Should render appropriate text when quantity is 3', () => {
      const wrapper = setUp({ productQuantity: 3 });
      expect(wrapper.find('.quantity-number').text()).toEqual('of 3 pieces');
    });
  });

  describe('Check how buttons reacts on click', () => {
    it('Should buttons call onSetModal()', () => {
      const onSetModalFn = jest.fn();
      const wrapper = setUp({ userProfile: undefined, onSetModal: onSetModalFn });
      const addToCartBtn = wrapper.find(Button).first();
      const buyBtn = wrapper.find(Button).last();
      addToCartBtn.simulate('click');
      buyBtn.simulate('click');
      expect(onSetModalFn.mock.calls).toHaveLength(2);
    });
    it('Should NOT buttons call onSetModal()', () => {
      const onSetModalFn = jest.fn();
      const wrapper = setUp({ onSetModal: onSetModalFn });
      const addToCartBtn = wrapper.find(Button).first();
      const buyBtn = wrapper.find(Button).last();
      addToCartBtn.simulate('click');
      buyBtn.simulate('click');
      expect(onSetModalFn.mock.calls).toHaveLength(0);
    });
  });
});
