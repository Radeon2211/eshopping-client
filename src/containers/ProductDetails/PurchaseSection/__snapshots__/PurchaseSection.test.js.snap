// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PurchaseSection /> Check how renders Snapshots should render everything correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ jLAphW"
  >
    <div
      class="sc-iqHYGH lcqKto"
      data-testid="PurchaseSection-choose-quantity-wrapper"
      direction="row"
      spacing=""
      width="100%"
      wrap="nowrap"
    >
      <div
        class="sc-fubCfw bLlOpR"
      >
        <button
          class="button minus"
          data-testid="ChooseQuantity-minus-btn"
          disabled=""
          type="button"
        >
          <div
            class="sc-iBPRYJ bwYaRv"
            color="#DADFE1"
            data-testid="ChooseQuantity-minus-icon"
          >
            <svg>
              minus.svg
            </svg>
          </div>
        </button>
        <input
          class="sc-gKsewC hNWjI"
          data-testid="NumberInput-quantity"
          name="quantity"
          type="number"
          value="1"
        />
        <button
          class="button plus"
          data-testid="ChooseQuantity-plus-btn"
          type="button"
        >
          <div
            class="sc-iBPRYJ ftEqkT"
            color=""
            data-testid="ChooseQuantity-plus-icon"
          >
            <svg>
              plus.svg
            </svg>
          </div>
        </button>
      </div>
      <span
        class="sc-eCssSg csvHNO"
        color=""
        data-testid="PurchaseSection-product-quantity"
        display="inline-block"
        overflow="visible"
        size="2"
        spacing="0px"
        transform=""
      >
        of 5 pieces
        <span
          class="sc-eCssSg htYuXx"
          color="#777"
          data-testid="PurchaseSection-quantity-in-cart"
          display="inline-block"
          overflow="visible"
          spacing="0px"
          transform=""
        >
          Â (3 in cart)
        </span>
      </span>
    </div>
    <button
      class="sc-gsTCUz jyQmMe"
      color="blue"
      type="button"
    >
      add to cart
    </button>
    <button
      class="sc-gsTCUz jyQmMe"
      color="blue"
      type="button"
    >
      buy now
    </button>
  </div>
</DocumentFragment>
`;

exports[`<PurchaseSection /> Check how renders Snapshots should render info that user is a seller 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ jLAphW"
  >
    <span
      class="sc-eCssSg hvnVVm"
      color=""
      display="inline-block"
      overflow="visible"
      size="2"
      spacing="0px"
      transform=""
    >
      Quantity: 5
    </span>
    <h4
      class="sc-kEjbxe EZwUf"
    >
      You are the seller of this product
    </h4>
  </div>
</DocumentFragment>
`;
